feat: Improve cluster node discovery with TLS and auth

waddles opened the pull request:
This commit enhances the service discovery for Redis/Valkey clusters, particularly for those using TLS and authentication.

Key changes:
1. The exporter now caches authentication credentials for dynamically discovered cluster nodes, allowing it to scrape metrics from clusters that require passwords.
2. The `/discover-cluster-nodes` endpoint now correctly preserves the `rediss://` or `valkeys://` scheme for TLS-enabled clusters.
3. The discovery handler can now connect to an arbitrary cluster specified by the `target` parameter, making it more flexible.
4. TLS configuration now includes the server name (SNI), improving security and compatibility.

A comprehensive set of tests has been added to validate these new capabilities, including scenarios for:
- Discovering and scraping nodes in password-protected clusters.
- Discovering and scraping nodes in TLS-enabled clusters.
- Handling different URI schemes (`redis://`, `rediss://`, `valkey://`, `valkeys://`).

Note: Depends on https://github.com/oliver006/docker-valkey-cluster/pull/1 to update the `valkey-cluster.tmpl` file used by envsubst when building the test cluster.

Commit:
feat: Improve cluster discovery with TLS and auth

This commit enhances the service discovery for Redis/Valkey clusters, particularly for those using TLS and authentication.

Key changes:
- The exporter now caches authentication credentials for dynamically discovered cluster nodes, allowing it to scrape metrics from clusters that require passwords.
- The `/discover-cluster-nodes` endpoint now correctly preserves the `rediss://` or `valkeys://` scheme for TLS-enabled clusters.
- The discovery handler can now connect to an arbitrary cluster specified by the `target` parameter, making it more flexible.
- TLS configuration now includes the server name (SNI), improving security and compatibility.

A comprehensive set of tests has been added to validate these new capabilities, including scenarios for:
- Discovering and scraping nodes in password-protected clusters.
- Discovering and scraping nodes in TLS-enabled clusters.
- Handling different URI schemes (`redis://`, `rediss://`, `valkey://`, `valkeys://`).

Commit:
docs: update service discovery with target param

oliver006 commented:
Thanks for the PR - I'll try to get to reviewing it in the next couple of days.

> Note: Depends on https://github.com/oliver006/docker-valkey-cluster/pull/1 to update the valkey-cluster.tmpl file used by envsubst when building the test cluster.

I just merged the `docker-valkey-cluster` PR and released tags 8.1.4 and 9.0.0 so you should be able to pull those in.

Commit:
chore: bump version and fix command options

oliver006 committed:
oliver006/valkey-cluster:8.1.5

oliver006 commented:
> I released 8.1.5 and 9.0.1 with your changes - and I bumped the version here to use oliver006/valkey-cluster:8.1.5

@waddles - a whole bunch of the tests related to Discovery failed, can you have a look?

waddles commented on main.go:
```diff
@@ -317,7 +318,11 @@ func main() {
+	if err != nil {
+		log.Fatalf("Couldn't parse redis addr, err: %s", err)
+	}
+	_, err = exp.CreateClientTLSConfig(u.Hostname())
```
ok so the tests all succeed but when I run the redis_exporter binary after running `make test-certs docker-env-up` then curl the exporter for cluster discovery I get
```
ERRO[81828] Cluster refresh failed: redisc: all nodes failed
tls: failed to verify certificate: x509: certificate relies on legacy Common Name field, use SANs instead
ERRO[81881] Cluster refresh failed: redisc: all nodes failed
tls: failed to verify certificate: x509: cannot validate certificate for 127.0.0.1 because it doesn't contain any IP SANs
```

The fix for that is to generate the test cert with SANs using a named pipe
```diff
diff --git a/contrib/tls/gen-test-certs.sh b/contrib/tls/gen-test-certs.sh
index 4a2810a..dafbcc8 100755
--- a/contrib/tls/gen-test-certs.sh
+++ b/contrib/tls/gen-test-certs.sh
@@ -29,4 +29,5 @@ openssl req \
         -CAserial ${dir}/ca.txt \
         -CAcreateserial \
         -days 3650 \
+        -extfile <(printf "subjectAltName=DNS:localhost,IP:127.0.0.1") \
         -out ${dir}/redis.crt

```
but I would be happy to revert main.go for my PR and let you play with SANs separately.

oliver006 commented:
@waddles - wanna see if we can break this down into smaller piece and get this moving forward one by one? 

Maybe let's start with (4) - the SNI part (it came up in #1125) - do you think you can break that part out as a separate PR and we'll get that merged and then we can look at the rest of the changes? 

(2) also seems useful and could be a next step, not sure yet about (1) and (3) - need to think about those more.

waddles commented:
I have a refactor of the TLS part that's almost ready to go, I just needed to finish testing it and have been flat out for months. I will try and clean it up against master and finish testing this weekend.

I really need this PR merged soon, I just finished a heap of elasticache upgrades and even though they added a bunch of nice new cloudwatch metrics, having independent prometheus metrics will allow us to make smarter dashboards and alerts, plus it give us something to compare CW against.

oliver006 commented:
> I have a refactor of the TLS part that's almost ready to go, I just needed to finish testing it and have been flat out for months. I will try and clean it up against master and finish testing this weekend.

Awesome, keep me posted.

Commit:
fix: make TLS SNI per-target, harden cluster discovery

Refines the cluster-discovery + TLS/auth feature: SNI is now a per-target
/scrape parameter rather than a global flag, plus several bugs found in
review are fixed.

SNI:
- CreateClientTLSConfig() reads Options.TlsServerName instead of taking a
  serverName argument, so ServerName can vary per scrape target.
- /scrape accepts a `tls-server-name` query parameter; when unset the SNI
  defaults to the target host (filled by redigo), so each discovered cluster
  node validates against its own certificate.
- Remove the global --tls-server-name flag and REDIS_EXPORTER_TLS_SERVER_NAME
  env var: one global SNI cannot validate multiple node hostnames, and the
  name collided with the server-side --tls-server-* flags.

Fixes:
- discoverClusterNodesHandler now preserves the TLS scheme on the no-target
  path by deriving it from e.redisAddr; previously a rediss:// cluster emitted
  redis:// node targets, silently downgrading scrapes to plaintext.
- Discovered-node passwords are cached under the same normalised key the
  lookup uses, so credentials resolve when --redis-user injects a username.
- startupNodeFromURI returns the url.Parse error instead of discarding it and
  dereferencing a nil *url.URL on a malformed target.
- The discovered-node password cache is invalidated on /-/reload so rotated
  credentials are not served from stale entries.

Cleanup:
- Fold connectToRedisCluster into connectToRedisClusterWithURI and extract
  schemeIsTLS / schemeFromURI / startupNodeFromURI / canonicalPasswordKey,
  removing duplicated scheme handling and a dead scheme-less branch.

Tests / infra:
- getTestExporter[WithOptions] take *testing.T and t.Skip when TEST_REDIS_URI
  is unset instead of panicking.
- Rewrite TestGetConfigForClientFunc with stdlib assertions, dropping the
  testify dependency.
- gen-test-certs.sh adds a DNS:localhost SAN; bump valkey-cluster test images
  to 9.0.1.

waddles commented:
Rebased my changes onto current master.

Tested on m4 mac (arm64) where possible (some older images are x86 only). Also tested on x86 linux - unfortunately valkey-bundle:8 (valkey 8.1.8) is compiled for x86-64-v3 / AVX2 but my 2013 cpu only has AVX so it dies with a SIGILL. I also notice the `docker-test` target fails but did not commit a fix for that.

Also fixed some minor bugs and dropped the `stretchr/testify` module to simplify go mod as per commit message.

oliver006 commented:
This is still a massive PR with many different changes lumped together - can you please break this down into a PR per feature? See my earlier comment, maybe start with item (4) - the TLS server name and then take it from there.

oliver006 commented:
Closing this.
I added tsl server name support in #1135 
Please add a new PR if you want to add a fix for (2)

Regarding (1) - why cache passwords in the exporter? Doesn't this expose us to the risk of stale passwords? What does get me? 

And regarding (3) - I think this could be helpful but if we add this then it needs safeguards and a flag cause otherwise it can be used to probe nodes (see the flag that was added to /scrape).

If you intend to add any of these features then please separate them by PR to make review easier.

waddles commented:
I'm not sure it's any clearer but I split it out into #1136, #1137 and #1138 (depends on 1137). Maybe doing that made the motivation for the change clearer in the PR body but I don't think there is much point to 1137 without 1138 as you would be able to discover the nodes properly but wouldn't be able to authenticate to them.

However, it did allow me find a couple of race conditions that were previously not considered.
